File: algospot/lan.cpp
#include <iostream>
#include <vector>
#include <math.h>
#include <iomanip>

using namespace std;

int main() {
#ifdef _DEBUG
	freopen("lan.in", "r", stdin);
#endif
	using FP = double;
	using pi = pair<int, int>;
	int c, n, m, a, b;
	const int MAX_V = 500, inf = 987654321;
	FP adj[MAX_V][MAX_V];
	pi pos[MAX_V];
	cin >> c;
	while (c--) {
		cin >> n >> m;
		for (int i = 0; i < n; i++) cin >> pos[i].first;
		for (int i = 0; i < n; i++) cin >> pos[i].second;
		for (int i = 0; i < n; i++)
			for (int j = 0; j < n; j++) {
				auto dx = (pos[i].first - pos[j].first);
				auto dy = (pos[i].second - pos[j].second);
				adj[i][j] = adj[j][i] = dx*dx + dy*dy;
			}
		while (m--) {
			cin >> a >> b;
			adj[a][b] = adj[b][a] = 0.0;
		}
		vector<bool> added(n, false);
		vector<FP> minWeight(n, 1e10);
		minWeight[0] = 0;
		FP dist = 0.0;
		for (int iter = 0; iter < n; iter++) {
			int u = -1;
			for (int i = 0; i < n; i++) {
				if (!added[i] && (u == -1 || minWeight[u] > minWeight[i])) 
					u = i;
			}
			added[u] = true;
			dist += sqrt(minWeight[u]);
			for (int i = 0; i < n; i++) {
				if (!added[i] && adj[u][i] < minWeight[i]) {
					minWeight[i] = adj[u][i];
				}
			}
		}
		cout << fixed << setprecision(9) << dist << endl;
	}
	return 0;
}

File: algospot/lan.py
from itertools import combinations
import sys


input = sys.stdin.readline

for cese in range(int(input())):
    n, m = map(int, input().split())
    x = list(map(int, input().split()))
    y = list(map(int, input().split()))
    adj = [[0]*n for _ in range(n)]
    for i, j in combinations(range(n), 2):
        adj[i][j] = adj[j][i] = (x[i]-x[j])**2 + (y[i]-y[j])**2
    for _ in range(m):
        a, b = map(int, input().split())
        adj[a][b] = adj[b][a] = 0
    minWeight = [1e10]*n
    dist = minWeight[0] = 0.0
    added = [False]*n
    for _ in range(n):
        u = -1
        for i in range(n):
            if not added[i] and (u == -1 or minWeight[u] > minWeight[i]):
                u = i
        added[u] = True
        dist += minWeight[u] ** 0.5
        for i in range(n):
            if not added[i] and adj[u][i] < minWeight[i]:
                minWeight[i] = adj[u][i]
    print(dist)
